packages/mendel-pipeline/src/step/gst.js

```diff
@@ -0,0 +1,174 @@
+        const currentGstConfig = this._transforms.find(({id}) => id === this._gsts[this._curGstIndex]);
+        const currentGst = require(currentGstConfig.plugin);
+
+        // If no GST is planned for this type, abort.
+        // If plugin doesn't want to deal with it, abort.
+        if (this._processed.has(entry.id) || !nextTransformIds.length || !currentGst.predicate(proxy)) {
+            return this.gstDone(entry);
+        }
+
+        const graph = this._registry.getDependencyGraph(entry.normalizedId, (depEntry) => {
+            const xformIds = this._registry.getClosestPlanTransformIds(depEntry.id, prevExec);
+            const dependecyMap = depEntry.getDependency(xformIds);
+            return Object.keys(dependecyMap).map(literal => {
+                // TODO remove the dep property
+                return dependecyMap[literal].main;
+            });
+        });
+
+        this.explorePermutation(graph, (chain, variation) => {
+            const [main] = chain;
+            this._processed.add(main.id);
+
+            // We need to create proxy for limiting API surface for plugin writers.
+            const context = this.getContext(nextTransformIds);
+            const chainProxy = chain.map(dep => {
+                const transformIds = this._registry.getClosestPlanTransformIds(dep.id, prevExec);
+                return EntryProxy.getFromEntry(dep, transformIds);
+            });
+            const result = currentGst.transform(context, chainProxy);
+            const [proxiedMain] = chainProxy;
+            proxiedMain.id = convertVariationalPath(
+                this._variations,
+                proxiedMain.id,
+                variation
+            );
+
+            if (result && result.source) {
+                if (entry.variation === variation) {
+                    this._registry.addTransformedSource({
+                        filePath: entry.id,
+                        transformIds: nextTransformIds,
+                        source: result.source,
+                        deps: result.deps,
+                    });
+                } else {
+                    context.addVirtualEntry({
+                        id: proxiedMain.id,
+                        originatingEntry: entry,
+                        source: result.source,
+                        deps: result.deps,
+                    });
+                }
+            }
+            this.gstDone(entry);
+        });
+    }
+
+    explorePermutation(graph, onPermutation) {
+        const configVariations = new Set();
+        this._variations.forEach(variation => configVariations.add(variation.id));
+
+        // graph has array of arrays. First, make a pass and gather all variation info
+        const variations = new Set();
+        graph.forEach(nodes => {
+            nodes.forEach(node => {
+                variations.add(node.variation);
+            });
+        });
+
+        Array.from(variations.keys())
```

Comment:
Yeah, I can refactor later, we should pair on this so I share how I would do it.